### PR TITLE
Treat decentralized DNS servers as available before first probe is completed

### PR DESCRIPTION
**Pull request by yrliou**
Resolves https://github.com/brave/brave-browser/issues/14513

This PR handles the case that DNS resolution failed because of falling back to non-secure DNS servers before the first probe of DoH server is completed and marked as available to use.

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: chromium_src/net/dns/host_resolver.cc [hunk 1]
```diff
@@ -0,0 +1,10 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/net/dns/brave_resolve_context.h"
+#include "net/dns/context_host_resolver.h"
+
+#define ResolveContext BraveResolveContext
+#include "../../../../net/dns/host_resolver.cc"
```
Comment (mkarolin) [inline chromium_src/net/dns/host_resolver.cc:10]: Nit: add #undef

File: chromium_src/net/dns/resolve_context.h [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+#ifndef BRAVE_CHROMIUM_SRC_NET_DNS_RESOLVE_CONTEXT_H_
+#define BRAVE_CHROMIUM_SRC_NET_DNS_RESOLVE_CONTEXT_H_
+
+namespace net {
+class BraveResolveContext;
+}  // namespace net
+
+#define GetDohServerAvailability virtual GetDohServerAvailability
+#define NumAvailableDohServers virtual NumAvailableDohServers
+#define BRAVE_RESOLVE_CONTEXT_H friend class BraveResolveContext;
```
Comment (mkarolin) [inline chromium_src/net/dns/resolve_context.h:15]: Nit: could follow the format:
<pre>
#define BRAVE_RESOLVE_CONTEXT_H\
 private: \
  friend class BraveResolveContext;
</pre>

**Review by mkarolin — APPROVED** [review]
`chromium_src` and `patches` changes LGTM

File: chromium_src/net/dns/host_resolver.cc (hunk 1 above)
Comment (yrliou) [reply chromium_src/net/dns/host_resolver.cc:10]: Added.

File: chromium_src/net/dns/resolve_context.h (hunk 2 above)
Comment (yrliou) [reply chromium_src/net/dns/resolve_context.h]: Fixed in https://github.com/brave/brave-core/pull/8212/commits/b20ad6265a782c561b51e0c52e4d13ff98846016.

**Review by bbondy — APPROVED** [review]

**Comment by yrliou** [comment]
post-init has unrelated audit-dep failure

**Pull request merged**